impl/src/planning.rs

```diff
@@ -416,82 +496,122 @@ impl<'planner> WorkspaceSubplanner<'planner> {
 
   /** Produces a crate context for each declared crate and dependency. */
   fn produce_crate_contexts(&self) -> Result<Vec<CrateContext>> {
+    // Build a list of all catalogs we want the context for
+    let mut catalogs = vec![self.crate_catalog];
+    for catalog in self.binary_dependencies.iter() {
+      catalogs.push(catalog);
+    }
+
+    // Build a list of all workspace files
+    let mut files: Vec<&CargoWorkspaceFiles> =
+      self.binary_deps_files
+      .iter()
+      .map(|(_key, val)| val)
+      .collect();
+    files.push(self.files);
+
     // Gather the checksums for all packages in the lockfile
     // which have them.
     //
     // Store the representation of the package as a tuple
     // of (name, version) -> checksum.
     let mut package_to_checksum = HashMap::new();
-    if let Some(lock_path) = self.files.lock_path_opt.as_ref() {
-      let lockfile = Lockfile::load(lock_path.as_path())?;
-      for package in lockfile.packages {
-        if let Some(checksum) = package.checksum {
-          package_to_checksum.insert(
-            (package.name.to_string(), package.version),
-            checksum.to_string(),
-          );
+    for workspace_files in files.iter() {
+      if let Some(lock_path) = workspace_files.lock_path_opt.as_ref() {
+        let lockfile = Lockfile::load(lock_path.as_path())?;
+        for package in lockfile.packages {
+          if let Some(checksum) = package.checksum {
+            package_to_checksum.insert(
+              (package.name.to_string(), package.version),
+              checksum.to_string(),
+            );
+          }
         }
       }
     }
 
-    self
-      .metadata
-      .resolve
-      .as_ref()
-      .ok_or_else(|| RazeError::Generic("Missing resolve graph".into()))?
-      .nodes
+    // Additionally, the binary dependencies need to have their checksums added as well
+    for (pkg, info) in self.settings.binaries.iter() {
+      let version = cargo_lock::Version::parse(info.req())?;
+      package_to_checksum.insert(
+        (pkg.clone(), version.clone()),
+        fetch_crate_checksum(&self.settings.index_url, pkg, &version.to_string())?,
+      );
+    }
+
+    let contexts = catalogs
       .iter()
-      .sorted_by_key(|n| &n.id)
-      .filter_map(|node| {
-        // UNWRAP: Node packages guaranteed to exist by guard in `produce_planned_build`
-        let own_crate_catalog_entry = self.crate_catalog.entry_for_package_id(&node.id).unwrap();
-        let own_package = own_crate_catalog_entry.package();
-
-        let checksum_opt =
-          package_to_checksum.get(&(own_package.name.clone(), own_package.version.clone()));
-
-        // Skip the root package (which is probably a junk package, by convention)
-        if own_crate_catalog_entry.is_root() {
-          return None;
-        }
+      .map(|catalog| {
+        (*catalog)
```

Comment:
Can you refactor this to be more imperative? Create helper functions, use variables, make sure things are clear.

I feel kind of strongly that this is an abomination of function chaining.

(not strongly enough to really push back if you push back, but man this looks gross to me and I think it will be hard for future maintainers to deal with)

Reply:
I'm happy you mentioned this because I feel `planning.rs` has kinda grown to be an abomination as well 😬 

I'll gladly look into this but would also love to see this file get split into a module 🤞

Reply:
Actually, I would rather leave this and spend some time first splitting `plannint.rs` into a module. I think the lines have become rather blurred between the internal components of this file. I've tried this already and have some good ideas on how to handle the unique situations.